File: mobile/lib/features/insights/domain/spending_analysis.dart
class CategorySpending {
  const CategorySpending({
    required this.categoryId,
    required this.categoryName,
    required this.currentSpent,
    required this.previousSpent,
    required this.changePercent,
  });

  final String categoryId;
  final String categoryName;
  final double currentSpent;
  final double previousSpent;
  final double changePercent;

  factory CategorySpending.fromJson(Map<String, dynamic> json) {
    return CategorySpending(
      categoryId: json['categoryId'] as String,
      categoryName: json['categoryName'] as String,
      currentSpent: (json['currentSpent'] as num).toDouble(),
      previousSpent: (json['previousSpent'] as num).toDouble(),
      changePercent: (json['changePercent'] as num).toDouble(),
    );
  }
}

class SpendingAnomaly {
  const SpendingAnomaly({
    required this.categoryId,
    required this.categoryName,
    required this.currentSpent,
    required this.previousSpent,
    required this.changePercent,
  });

  final String categoryId;
  final String categoryName;
  final double currentSpent;
  final double previousSpent;
  final double changePercent;

  factory SpendingAnomaly.fromJson(Map<String, dynamic> json) {
    return SpendingAnomaly(
      categoryId: json['categoryId'] as String,
      categoryName: json['categoryName'] as String,
      currentSpent: (json['currentSpent'] as num).toDouble(),
      previousSpent: (json['previousSpent'] as num).toDouble(),
      changePercent: (json['changePercent'] as num).toDouble(),
    );
  }
}

class TopCategory {
  const TopCategory({
    required this.categoryId,
    required this.categoryName,
    required this.amount,
    required this.percentOfTotal,
  });

  final String categoryId;
  final String categoryName;
  final double amount;
  final double percentOfTotal;

  factory TopCategory.fromJson(Map<String, dynamic> json) {
    return TopCategory(
      categoryId: json['categoryId'] as String,
      categoryName: json['categoryName'] as String,
      amount: (json['amount'] as num).toDouble(),
      percentOfTotal: (json['percentOfTotal'] as num).toDouble(),
    );
  }
}

class SpendingAnalysis {
  const SpendingAnalysis({
    required this.householdId,
    required this.period,
    required this.periodStartUtc,
    required this.periodEndUtc,
    required this.totalSpent,
    required this.previousPeriodTotalSpent,
    required this.spendingChangePercent,
    required this.categories,
    required this.anomalies,
    required this.topCategories,
  });

  final String householdId;
  final String period;
  final String periodStartUtc;
  final String periodEndUtc;
  final double totalSpent;
  final double previousPeriodTotalSpent;
  final double spendingChangePercent;
  final List<CategorySpending> categories;
  final List<SpendingAnomaly> anomalies;
  final List<TopCategory> topCategories;

  factory SpendingAnalysis.fromJson(Map<String, dynamic> json) {
    return SpendingAnalysis(
      householdId: json['householdId'] as String,
      period: json['period'] as String,
      periodStartUtc: json['periodStartUtc'] as String,
      periodEndUtc: json['periodEndUtc'] as String,
      totalSpent: (json['totalSpent'] as num).toDouble(),
      previousPeriodTotalSpent:
          (json['previousPeriodTotalSpent'] as num).toDouble(),
      spendingChangePercent:
          (json['spendingChangePercent'] as num).toDouble(),
      categories: (json['categories'] as List<dynamic>)
          .map((e) => CategorySpending.fromJson(e as Map<String, dynamic>))
          .toList(),
      anomalies: (json['anomalies'] as List<dynamic>)
          .map((e) => SpendingAnomaly.fromJson(e as Map<String, dynamic>))
          .toList(),
      topCategories: (json['topCategories'] as List<dynamic>)
          .map((e) => TopCategory.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: backend/src/Modules/Insights/Domain/SpendingAnalysis.cs
namespace MoneyTracker.Modules.Insights.Domain;

public sealed record CategorySpending(
    Guid CategoryId,
    string CategoryName,
    decimal CurrentSpent,
    decimal PreviousSpent,
    double ChangePercent);

public sealed record TopCategory(
    Guid CategoryId,
    string CategoryName,
    decimal Amount,
    double PercentOfTotal);

public sealed class SpendingAnalysis
{
    public decimal TotalSpent { get; }
    public decimal PreviousPeriodTotalSpent { get; }
    public double SpendingChangePercent { get; }
    public IReadOnlyCollection<CategorySpending> Categories { get; }
    public IReadOnlyCollection<TopCategory> TopCategories { get; }
    public IReadOnlyCollection<SpendingAnomaly> Anomalies { get; }

    private SpendingAnalysis(
        decimal totalSpent,
        decimal previousPeriodTotalSpent,
        double spendingChangePercent,
        IReadOnlyCollection<CategorySpending> categories,
        IReadOnlyCollection<TopCategory> topCategories,
        IReadOnlyCollection<SpendingAnomaly> anomalies)
    {
        TotalSpent = totalSpent;
        PreviousPeriodTotalSpent = previousPeriodTotalSpent;
        SpendingChangePercent = spendingChangePercent;
        Categories = categories;
        TopCategories = topCategories;
        Anomalies = anomalies;
    }

    public static SpendingAnalysis Compute(
        IReadOnlyCollection<CategorySpendingData> currentPeriodData,
        IReadOnlyCollection<CategorySpendingData> previousPeriodData,
        double anomalyThresholdPercent)
    {
        var previousLookup = previousPeriodData
            .ToDictionary(d => d.CategoryId, d => d.TotalSpent);

        var totalSpent = currentPeriodData.Sum(d => d.TotalSpent);
        var previousTotalSpent = previousPeriodData.Sum(d => d.TotalSpent);
        var overallChange = previousTotalSpent == 0
            ? 0.0
            : (double)((totalSpent - previousTotalSpent) / previousTotalSpent) * 100;

        var categories = new List<CategorySpending>();
        var anomalies = new List<SpendingAnomaly>();

        foreach (var current in currentPeriodData)
        {
            var previousSpent = previousLookup.GetValueOrDefault(current.CategoryId, 0m);
            var changePercent = previousSpent == 0
                ? 0.0
                : (double)((current.TotalSpent - previousSpent) / previousSpent) * 100;

            categories.Add(new CategorySpending(
                current.CategoryId,
                current.CategoryName,
                current.TotalSpent,
                previousSpent,
                changePercent));

            var anomaly = SpendingAnomaly.Detect(
                current.CategoryId,
                current.CategoryName,
                current.TotalSpent,
                previousSpent,
                anomalyThresholdPercent);

            if (anomaly is not null)
            {
                anomalies.Add(anomaly);
            }
        }

        var topCategories = currentPeriodData
            .OrderByDescending(d => d.TotalSpent)
            .Select(d => new TopCategory(
                d.CategoryId,
                d.CategoryName,
                d.TotalSpent,
                totalSpent == 0 ? 0.0 : (double)(d.TotalSpent / totalSpent) * 100))
            .ToArray();

        return new SpendingAnalysis(
            totalSpent,
            previousTotalSpent,
            overallChange,
            categories.AsReadOnly(),
            topCategories,
            anomalies.AsReadOnly());
    }
}

public sealed record CategorySpendingData(
    Guid CategoryId,
    string CategoryName,
    decimal TotalSpent);
